Improve Amazon S3 client guide (#135)
wso2/ballerina-integrator: examples/saas-clients/amazons3/working-with-amazons3-client/README.md

# Working with Amazon S3 client

## Overview

> The Ballerina Amazon S3 client provides connectivity to Amazon Simple Storage Service API. It allows users to interact 

Reviewer: ```suggestion
The Ballerina Amazon S3 client enables you to connect to the Amazon Simple Storage Service API. It allows users to interact 
```

Author: Fixed.

> with Amazon S3 API to create, store and download and use data with other services. 

Reviewer: ```suggestion
with the Amazon S3 API to create, store, download, and use data with other services. 
```

Generally it is best to use a serial comma (comma before and) in lists.

Author: Fixed.

> This example explains how to use the S3 client to connect with Amazon S3 instance and perform the following operations:

Reviewer: ```suggestion
This example explains how to use the S3 client to connect with the Amazon S3 instance and perform the following operations:
```

Author: Fixed.

* Create Bucket
* Create Object
* Get Object
* Delete Object
* Delete Bucket

## Prerequisites

> - [Ballerina Distribution](https://ballerina.io/learn/getting-started/)

Reviewer: Any reason why we are not pointing directly to https://ballerina.io/downloads/?

Author: Isn't it better go provide step by step guidelines page for the user to download and install ballerina rather than directly providing with the downloads page?

- A Text Editor or an IDE 
> **Tip**: For a better development experience, install one of the following Ballerina IDE plugins: [VSCode](https://marketplace.visualstudio.com/items?itemName=ballerina.ballerina), [IntelliJ IDEA](https://plugins.jetbrains.com/plugin/9520-ballerina)

> - Obtain the credentials such as Access Key, Secret Access Key (API Secret) from Amazon S3 by following the steps below:

Reviewer: ```suggestion
- Obtain the credentials such as Access Key and Secret Access Key (API Secret) from Amazon S3 by following the steps below:
```

Author: Fixed.

>     1. Create an amazon account by visiting <https://aws.amazon.com/s3/>

Reviewer: ```suggestion
    1. Create an Amazon S3 account by visiting <https://aws.amazon.com/s3/>
```

Author: Fixed.

>     2. Create a new access key, which includes a new secret access key.

Reviewer: Do you mean "new root access key" here?

        - To create a new secret access key for your root account, use the [security credentials](https://console.aws.amazon.com/iam/home?#security_credential) page. Expand the Access Keys section, and then click Create New Root Key.
        - To create a new secret access key for an IAM user, open the [IAM console](https://console.aws.amazon.com/iam/home?region=us-east-1#home). Click Users in the Details pane, click the appropriate IAM user, and then click Create Access Key on the Security Credentials tab.
    3. Download the newly created credentials, when prompted to do so in the key creation wizard.

Let's get started with a simple Ballerina program to integrate with Amazon S3.  

> ### Implementation

Reviewer: Can we come up with a rough diagram on what we build here?

Author: Added a diagram.

#### Creating the module structure
Ballerina is a complete programming language that can have any custom project structure as you wish. Although the 
language allows you to have any module structure, you can use the following simple module structure for this project.
```
working-with-amazons3-client
    └── working_with_amazons3_client.bal
```
- Create a directory called `working-with-amazons3-client` in your local machine and also create the empty .bal files.

> - Then open a terminal, navigate to `working-with-amazons3-client`, and run the Ballerina project initializing toolkit.

Reviewer: ```suggestion
- Open a terminal, navigate to `working-with-amazons3-client`, and run the Ballerina project initializing toolkit.
```

Author: Fixed.

amazons3:ClientConfiguration amazonS3Config = {
    accessKeyId: accessKeyId,
    secretAccessKey: secretAccessKey
};

@http:ServiceConfig {
    basePath: "/amazons3"
}
service amazonS3Service on new http:Listener(9090) {
    @http:ResourceConfig {
        methods: ["POST"],
        path: "/{bucketName}"
    }
    // Function to create a new bucket.
    resource function createBucket(http:Caller caller, http:Request request, string bucketName) {
        // Create AmazonS3 client with the above amazonS3Config.
        amazons3:AmazonS3Client|error amazonS3Client = new(amazonS3Config);     
        if (amazonS3Client is amazons3:AmazonS3Client) {
            // Define new response. 
            http:Response backendResponse = new();
            // Invoke createBucket remote function from amazonS3Client.
            error? response = amazonS3Client->createBucket(untaint bucketName);
            if (response is error) {
                // Send the error response.
                createAndSendErrorResponse(caller, untaint <string>response.detail().message, 
                                BUCKET_CREATION_ERROR_MSG);
            } else {

>                 // If there is no error, then bucket created successfully. Send the success response.

Reviewer: ```suggestion
                // If there is no error, the bucket is created successfully. Send the success response.
```

Author: Fixed.

        amazons3:AmazonS3Client|error amazonS3Client = new(amazonS3Config);
        if (amazonS3Client is amazons3:AmazonS3Client) {
            // Define new response. 
            http:Response backendResponse = new();
            // Extract the object content from request payload.
            string|xml|json|byte[]|error objectContent = extractRequestContent(request);
            if objectContent is error {
                // Send the error response.
                createAndSendErrorResponse(caller, untaint <string>objectContent.detail().message, 
                                PAYLOAD_EXTRACTION_ERROR_MSG);
            } else {
                // Invoke createObject remote function from amazonS3Client.
                error? response = amazonS3Client->createObject(untaint bucketName, untaint objectName,
                                                    untaint objectContent);
                if (response is error) {
                    // Send the error response.
                    createAndSendErrorResponse(caller, untaint <string>response.detail().message,
                                    OBJECT_CREATION_ERROR_MSG);
                } else {

>                     // If there is no error, then object created successfully. Send the success response.

Reviewer: ```suggestion
                    // If there is no error, the object is created successfully. Send the success response.
```

Author: Fixed.

                                 "Error while creating object on Amazon S3.");
            }
        } else {
            // Send the error response.
            createAndSendErrorResponse(caller, <string>amazonS3Client.detail().message, CLIENT_CREATION_ERROR_MSG);
        }
    }

    @http:ResourceConfig {
        methods: ["DELETE"],
        path: "/{bucketName}/{objectName}"
    }
    // Function to delete object.
    resource function deleteObject(http:Caller caller, http:Request request, string bucketName, string objectName) {
        // Create AmazonS3 client with the above amazonS3Config.
        amazons3:AmazonS3Client|error amazonS3Client = new(amazonS3Config);
        if (amazonS3Client is amazons3:AmazonS3Client) {
            // Define new response. 
            http:Response backendResponse = new();
            error? response = amazonS3Client->deleteObject(untaint bucketName, untaint objectName);
            if (response is error) {
                // Send the error response.
                createAndSendErrorResponse(caller, untaint <string>response.detail().message, 
                                        OBJECT_DELETION_ERROR_MSG);
            } else {

>                 // If there is no error, then object deleted successfully. Send the success response.

Reviewer: ```suggestion
                // If there is no error, the object is deleted successfully. Send the success response.
```

Author: Fixed.

            }
        } else {
            // Send the error response.
            createAndSendErrorResponse(caller, <string>amazonS3Client.detail().message, CLIENT_CREATION_ERROR_MSG);
        }
    }

    @http:ResourceConfig {
        methods: ["DELETE"],
        path: "/{bucketName}"
    }
    // Function to delete bucket.
    resource function deleteBucket(http:Caller caller, http:Request request, string bucketName, string objectName) {
        // Create AmazonS3 client with the above amazonS3Config.
        amazons3:AmazonS3Client|error amazonS3Client = new(amazonS3Config);
        if (amazonS3Client is amazons3:AmazonS3Client) {
            // Define new response. 
            http:Response backendResponse = new();
            // Invoke deleteBucket remote function from amazonS3Client.
            error? response = amazonS3Client->deleteBucket(untaint bucketName);
            if (response is error) {
                // Send the error response.
                createAndSendErrorResponse(caller, untaint <string>response.detail().message, 
                                        BUCKET_DELETION_ERROR_MSG);
            } else {

>                 // If there is no error, then bucket deleted successfully. Send the success response.

Reviewer: ```suggestion
                // If there is no error, the bucket is deleted successfully. Send the success response.
```

Author: Fixed.

                backendResponse.setTextPayload(untaint string `${bucketName} deleted from Amazon S3.`, 
                                        contentType = "text/plain");
                respondAndHandleError(caller, backendResponse, RESPOND_ERROR_MSG);
            }
        } else {
            // Send the error response.
            createAndSendErrorResponse(caller, <string>amazonS3Client.detail().message, CLIENT_CREATION_ERROR_MSG);
        }
    }
}

> // Function to extract the object content from request payload

Reviewer: Add more inline comments to this section. Users need to know what you are doing in each line.

    if (binaryObjectContent is byte[]) {
        return binaryObjectContent;
    } else {
        error err = error(ERROR_CODE, { message : INVALID_PAYLOAD_MSG });
        return err;
    }
}

// Function to create the error response.
function createAndSendErrorResponse(http:Caller caller, string errorMessage, string respondErrorMsg) {
    http:Response response = new;
    //Set 500 status code.
    response.statusCode = 500;
    //Set the error message to the error response payload.
    response.setPayload(<string> errorMessage);
    respondAndHandleError(caller, response, respondErrorMsg);
}

// Function to send the response back to the client and handle the error.
function respondAndHandleError(http:Caller caller, http:Response response, string respondErrorMsg) {
    // Send response to the caller.
    var respond = caller->respond(response);
    if (respond is error) {
        log:printError(respondErrorMsg, err = respond);
    }
}
```
### Deployment

#### Deploying locally
You can deploy the services that you developed above in your local environment. You can create the Ballerina executable archives (.balx) first as follows.

**Building**

> Navigate to working-with-amazons3-client and execute the following command.

Reviewer: ```suggestion
Navigate to `working-with-amazons3-client` and execute the following command.
```

Author: Fixed.

```bash
$ ballerina build working_with_amazons3_client.bal
```

After the build is successful, there will be a working_with_amazons3_client.balx file inside the target directory. You can execute it as follows.

```bash
$ ballerina run working_with_amazons3_client.balx
```

### Testing

- Navigate to `working-with-amazons3-client`, and execute the following command to start the service:

```bash
$ ballerina run working_with_amazons3_client.bal
```

#### Test createBucket service
- Invoke the following curl request to create a new bucket.
```
curl -v -X POST http://localhost:9090/amazons3/firstbalbucket
```
You will see the response as follows:
```
firstbalbucket created on Amazon S3.
```

#### Test createObject service
##### (I) JSON Content

> Create a file called content.json with following json content:

Reviewer: ```suggestion
Create a file called `content.json` with following JSON content:
```
Add all file names, method names, values, etc. within single backticks.

```json
{
    "name": "John Doe",
    "dob": "1940-03-19",
    "ssn": "234-23-525",
    "address": "California",
    "phone": "8770586755",
    "email": "johndoe@gmail.com",
    "doctor": "thomas collins",
    "hospital": "grand oak community hospital",
    "cardNo": "7844481124110331",
    "appointment_date": "2025-04-02"
}
```
- Invoke the following curl request to create a new object in the newly created bucket.
```
curl -v -X POST --data @content.json http://localhost:9090/amazons3/firstbalbucket/firstObject.json --header "Content-Type:application/json"
```
You will see the response as follows:
```
firstObject.json created on Amazon S3 bucket : firstbalbucket.
```

##### (II) Binary Content

> Let's upload an image (sample.jpg) to the s3 bucket we created above. 

Reviewer: Shall we add the link to sample.jpg?

Author: sample.jpg can be any image.

- Invoke the following curl request to create a new object in the newly created bucket.
```
curl -v -X POST http://localhost:9090/amazons3/firstbalbucket/image.jpg -H 'Content-Type: image/jpg' -T sample.jpg
```
You will see the response as follows:
```
image.jpg created on Amazon S3 bucket : firstbalbucket.
```

#### Test getObject service
##### (I) JSON Content

> - Set the responseContentType as application/json to retrieve a json object and invoke the following curl request to get the newly created object.

Reviewer: ```suggestion
- Set the `responseContentType` as `application/json` to retrieve a JSON object and invoke the following curl request to get the newly created object.
```

Author: Fixed.

```
curl -v -X GET http://localhost:9090/amazons3/firstbalbucket/firstObject.json?responseContentType=application/json
```
You will see the response as follows:
```json
{
    "name": "John Doe",
    "dob": "1940-03-19",
    "ssn": "234-23-525",
    "address": "California",
    "phone": "8770586755",
    "email": "johndoe@gmail.com",
    "doctor": "thomas collins",
    "hospital": "grand oak community hospital",
    "cardNo": "7844481124110331",
    "appointment_date": "2025-04-02"
}
```

##### (II) Binary Content
- Set the responseContentType as image/jpg and use following URL to open newly created image on browser.
```
http://localhost:9090/amazons3/firstbalbucket/image.jpg?responseContentType=image/jpg
```

- Set the responseContentType as application/octet-stream and use the following URL to download newly created image.
```
http://localhost:9090/amazons3/firstbalbucket/image.jpg?responseContentType=application/octet-stream
```

#### Test deleteObject service

> - Invoke the following curl request to delete above object.

Reviewer: ```suggestion
- Invoke the following curl request to delete the above object.
```

Author: Fixed.

```
curl -v -X DELETE http://localhost:9090/amazons3/firstbalbucket/firstObject.json
```
You will see the response as follows:
```
firstObject.json deleted on Amazon S3 bucket : firstbalbucket.
```

#### Test deleteBucket service

> - Invoke the following curl request to delete above bucket.

Reviewer: ```suggestion
- Invoke the following curl request to delete the above bucket.
```

Author: Fixed.

```
curl -v -X DELETE http://localhost:9090/amazons3/firstbalbucket
```

> You will see the response as follows:

Reviewer: ```suggestion
You see the response as follows:
```

Author: Fixed.

```
working-with-amazons3-client
    └── working_with_amazons3_client.bal
```
- Create a directory called `working-with-amazons3-client` in your local machine and also create the empty .bal files.
- Open a terminal, navigate to `working-with-amazons3-client`, and run the Ballerina project initializing toolkit.
```
$ ballerina init
```
In the working-with-amazons3-client directory where you have your sample, create a `ballerina.conf` file and add the  details you obtained above within the quotes.

```
ACCESS_KEY_ID = ""
SECRET_ACCESS_KEY = ""
```
Now that you have created the project structure, the next step is to develop the service.

#### Developing the service

The following code is the completed sample which exposes the following services:
- createBucket: Creates a new bucket on Amazon S3 instance with the provided name.
- createObject: Creates a new object on an existing Amazon S3 bucket.
- getObject: Retrieves an object from an Amazon S3 bucket.
- deleteObject: Deletes an existing object from an Amazon S3 bucket.
- deleteBucket: Deletes specified bucket from Amazon S3 instance.

> ```ballerina

Reviewer 2: Use syntax given in www README on how to add code segments and files.
